zebrad/src/components/inbound/tests.rs

```diff
@@ -171,13 +171,129 @@ async fn mempool_advertise_transaction_ids() -> Result<(), crate::BoxError> {
     Ok(())
 }
 
+#[tokio::test]
+async fn mempool_transaction_expiration() -> Result<(), crate::BoxError> {
+    // Get a block that has at least one non coinbase transaction
+    let block: Block = zebra_test::vectors::BLOCK_MAINNET_982681_BYTES.zcash_deserialize_into()?;
+
+    // Use the first transaction that is not coinbase to test expiration
+    let tx1 = &*(block.transactions[1]).clone();
+    let mut tx1_id = tx1.unmined_id();
+
+    // Change the expiration height of the transaction to block one
+    let mut tx1 = tx1.clone();
+    *tx1.expiry_height_mut() = zebra_chain::block::Height(1);
+
+    // Use the second transaction that is not coinbase to trigger `remove_expired_transactions()`
+    let tx2 = block.transactions[2].clone();
+    let mut tx2_id = tx2.unmined_id();
+
+    // Get services
+    let (inbound_service, _, mut tx_verifier, _peer_set, state_service) = setup(false).await;
+
+    // Push test transaction
+    let request = inbound_service
+        .clone()
+        .oneshot(Request::PushTransaction(tx1.clone().into()));
+    // Simulate a successful transaction verification
+    let verification = tx_verifier.expect_request_that(|_| true).map(|responder| {
+        tx1_id = responder.request().tx_id();
+        responder.respond(tx1_id);
+    });
+    let (response, _) = futures::join!(request, verification);
+    match response {
+        Ok(Response::Nil) => (),
+        _ => unreachable!("`PushTransaction` requests should always respond `Ok(Nil)`"),
+    };
+
+    // Use `Request::MempoolTransactionIds` to check the transaction was inserted to mempool
+    let request = inbound_service
+        .clone()
+        .oneshot(Request::MempoolTransactionIds)
+        .await;
+
+    match request {
+        Ok(Response::TransactionIds(response)) => {
+            assert_eq!(response, vec![tx1_id])
+        }
+        _ => unreachable!(
+            "`MempoolTransactionIds` requests should always respond `Ok(Vec<UnminedTxId>)`"
+        ),
+    };
+
+    // Add a new block to the state (make the chain tip advance)
+    let block_one: Arc<Block> = zebra_test::vectors::BLOCK_MAINNET_1_BYTES
+        .zcash_deserialize_into()
+        .unwrap();
+    state_service
+        .clone()
+        .oneshot(zebra_state::Request::CommitFinalizedBlock(
+            block_one.clone().into(),
+        ))
+        .await
+        .unwrap();
+
```

Comment:
Missing test cases:
- Can you check that `tx1` isn't expired after `block_one`? (it should only expire after block 2)
- Can you check that a transaction with no expiry height never expires?

I think that covers the consensus rules:
> Transactions will have a new field, nExpiryHeight, which will set the block height after which transactions will be removed from the mempool if they have not been mined.
> No limit: To set no limit on transactions (so that they do not expire), nExpiryHeight should be set to 0.

Reply:
https://github.com/ZcashFoundation/zebra/pull/2774/commits/d39d84626be1402d4167bcf1ad9fa4d513321922